example_logger_test.go

```diff
@@ -0,0 +1,146 @@
+			AbsPath: "./b/c", // malformed absolute filepath to simulate a "bug"
+		}, nil
+	} else if req.ID == "req3" {
+		return &Resource{
+			ID:      req.ID,
+			AbsPath: "/some/correct/path/data.json",
+		}, nil
+	}
+
+	return nil, eris.Wrapf(ErrNotFound, "failed to get resource '%v'", req.ID)
+}
+
+func GetRelPath(base string, path string) (string, error) {
+	relPath, err := filepath.Rel(base, path)
+	if err != nil {
+		// it's generally useful to wrap external errors with a type that you know how to handle
+		// first (e.g. ErrInternalServer). this will help if/when you want to do error inspection
+		// via eris.Is(err, ErrInternalServer) or eris.Cause(err).
+		return "", eris.Wrap(ErrInternalServer, err.Error())
+	}
+	return relPath, nil
+}
+
+type Response struct {
+	RelPath string
+}
+
+func ProcessResource(req Request) (*Response, error) {
+	if err := req.Validate(); err != nil {
+		// simply return the error if there's no additional context
+		return nil, err
+	}
+
+	resource, err := GetResource(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// do some processing on the data
+	relPath, err := GetRelPath("/rel/base/", resource.AbsPath)
+	if err != nil {
+		// wrap the error again if you want to add more context
+		return nil, eris.Wrapf(err, "failed to get relative path for resource '%v'", resource.ID)
+	}
+
+	return &Response{
+		RelPath: relPath,
+	}, nil
+}
+
+type LogReq struct {
+	Method string
+	Req    Request
+	Res    *Response
+	Err    error
+}
+
+func LogRequest(logger *logrus.Logger, logReq LogReq) {
+	fields := logrus.Fields{
+		"method":  logReq.Method,
+		"request": logReq.Req,
+	}
+	if logReq.Err != nil {
+		// it's generally a good idea to contain error formatting logic inside a utility method like
+		// this one to ensure that all errors are logged uniformly
+		unpackedErr := eris.Unpack(logReq.Err)
+		fields["error"] = unpackedErr.ToJSON(eris.NewDefaultFormat(true))
+		logger.WithFields(fields).Error("method completed with error")
+	} else {
+		fields["response"] = *logReq.Res
+		logger.WithFields(fields).Info("method completed successfully")
+	}
+}
+
+// This example demonstrates how to integrate eris with a JSON logger (e.g. logrus). It's broken
+// into several methods to show the formatted output for wrapped errors, and it includes three
+// failing cases to demonstrate how your error logs should look in different scenarios.
+func Example_logger() {
+	// Setup JSON logger
+	logger := logrus.New()
+	logger.SetOutput(os.Stdout)
+	logger.SetFormatter(&logrus.JSONFormatter{})
+
+	// Example requests
+	reqs := []Request{
+		{
+			ID: "", // bad request
+		},
+		{
+			ID: "res1", // not found
+		},
+		{
+			ID: "req2", // server error
+		},
+		{
+			ID: "req3", // success
+		},
+	}
+
+	// Process the example requests and log the results
+	for _, req := range reqs {
+		res, err := ProcessResource(req)
+		LogRequest(logger, LogReq{
+			Method: "ProcessResource",
+			Req:    req,
+			Res:    res,
+			Err:    err,
+		})
+	}
```

Comment:
should i add example output? it will be very long

Reply:
Yeah, i think the example itself is very long. We need to convey some parts in quick read